pkg/app/api/grpcapi/web_api.go

```diff
@@ -1350,52 +1344,160 @@ func (a *WebAPI) getInsightDataForDeployFrequency(ctx context.Context, projectID
 	}
 
 	for i := 0; i < int(req.DataPointCount); i++ {
-		target := movePoint(start, i)
-
-		filters := []datastore.ListFilter{
-			{
-				Field:    "ProjectId",
-				Operator: "==",
-				Value:    projectID,
-			},
-			{
-				Field:    "CreatedAt",
-				Operator: ">=",
-				Value:    target.Unix(),
-			},
-			{
-				Field:    "CreatedAt",
-				Operator: "<",
-				Value:    movePoint(target, 1).Unix(), // target's finish time on unix time
-			},
+		targetRangeFrom := movePoint(start, i)
+		targetRangeTo := movePoint(targetRangeFrom, 1)
+
+		switch req.MetricsKind {
+		case model.InsightMetricsKind_DEPLOYMENT_FREQUENCY:
+			count, err := a.getInsightDataForDeployFrequency(ctx, projectID, req.ApplicationId, targetRangeFrom, targetRangeTo)
+			if err != nil {
+				return nil, err
+			}
+			counts[i] = count
+		case model.InsightMetricsKind_CHANGE_FAILURE_RATE:
+			count, err := a.getInsightDataForChangeFailureRate(ctx, projectID, req.ApplicationId, targetRangeFrom, targetRangeTo)
+			if err != nil {
+				return nil, err
+			}
+			counts[i] = count
```

Comment:
nit: To avoid to be deeply nested, why not just do like:
```go
var getInsightData func(context.Context, string, string, time.Time, time.Time)
switch req.MetricsKind {
case model.InsightMetricsKind_DEPLOYMENT_FREQUENCY:
                   getInsightData = getInsightDataForDeployFrequency
case model.InsightMetricsKind_CHANGE_FAILURE_RATE:
                   getInsightData = getInsightDataForChangeFailureRate
...
```

Reply:
Great idea. Thanks :)